### PR TITLE
[MHA] Implement MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR API Class

### LINKED ISSUE #2831: [MHA] Implement MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR in Graph API
Implement the following.
- [x] Tests for miopen::graphapi::VariantPack and miopen::graphapi::VariantPackBuilder
- [x] Classes miopen::graphapi::VariantPack and miopen::graphapi::VariantPackBuilder
- [x] Tests for C functions for MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR
- [x] miopenBackendCreateDescriptor
- [x] miopenBackendSetAttribute
- [x] miopenBackendFinalize
- [x] miopenBackendGetAttribute

### PR DESCRIPTION
**Pull request by scerzh**
Issue [#2831](https://github.com/ROCm/MIOpen/issues/2831)

The following tasks were implemented
- [x] Tests for C functions for MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR
- [x] miopenBackendCreateDescriptor
- [x] miopenBackendSetAttribute
- [x] miopenBackendFinalize
- [x] miopenBackendGetAttribute

C API -> BackendVariantPackDescriptor -> VariantPackBuilder -> VariantPack

BackendVariantPackDescriptor, a polymorphic descriptor class, is utilized by C-API functions to handle create/set/finalize/get/execute requests.

Builder pattern is used for set and finalize API requests.

VariantPack class holds data pointers for non-virtual tensors and for workspace.

This PR closes #2831 

@atamazov , you are welcome to review

### CODE REVIEW

File: test/gtest/graphapi_variant_pack.cpp [hunk 1]
```diff
@@ -308,86 +284,170 @@ TEST_P(GraphApiVariantPackBuilder, RetrieveAttributes)
...
+        bool valid                                             = false;
+        std::tie(valid, mTensorIds, mDataPointers, mWorkspace) = GetParam();
+        execute.descriptor = {"MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR",
+                              MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR,
+                              valid,
+                              {&mTensorIds, &mDataPointers, &mWorkspace}};
+    }
+};
+
+TEST_P(GraphApiVariantPack, CFuncions) { execute(); }
```
Comment (amberhassaan) [inline test/gtest/graphapi_variant_pack.cpp]: minor: Typo s/Funcions/Functions/

File: test/gtest/graphapi_variant_pack.cpp [hunk 2]
```diff
@@ -308,86 +284,170 @@ TEST_P(GraphApiVariantPackBuilder, RetrieveAttributes)
...
+    }
+};
+
+} // namespace
+
+class GraphApiVariantPack : public testing::TestWithParam<GraphApiVariantPackTuple>
+{
+protected:
+    // to be used in the test
+    GTestGraphApiExecute<GTestDescriptorAttribute*> execute;
```
Comment (amberhassaan) [inline test/gtest/graphapi_variant_pack.cpp]: question: how is this member `execute` initialized?

**Review by amberhassaan — APPROVED** [review]
Looks good.

File: test/gtest/graphapi_variant_pack.cpp (hunk 2 above)
Comment (scerzh) [reply test/gtest/graphapi_variant_pack.cpp]: With "empty" default values as follows:
```C++
template <typename AttributePointer = std::shared_ptr<GTestDescriptorAttribute>>
struct GTestDescriptor
{
    const char* textName               = "";
    miopenBackendDescriptorType_t type = miopenBackendDescriptorType_t(0);
    bool attrsValid                    = false;
    std::vector<AttributePointer> attributes;
};

template <typename AttributePointer = std::shared_ptr<GTestDescriptorAttribute>>
struct GTestGraphApiExecute
{
    GTestDescriptor<AttributePointer> descriptor;
```
SetUp() should set values according to passed parameters.

File: test/gtest/graphapi_variant_pack.cpp (hunk 1 above)
Comment (scerzh) [reply test/gtest/graphapi_variant_pack.cpp]: done

**Comment by scerzh** [comment]
@junliume could we merge this one?

**Review by junliume — APPROVED** [review]

**Pull request merged**